backend/session/direct-ipc.c

```diff
@@ -36,8 +42,12 @@ static bool have_permissions(void) {
 #else
 static bool have_permissions(void) {
 	if (geteuid() != 0) {
+#ifdef __FreeBSD__
+		wlr_log(L_INFO, "Do not have root privileges; make sure you have write access to /dev/ttyv*, /dev/drm/* and /dev/input/*");
```

Comment:
How about testing for access instead? Can do this on both platforms.

Reply:
Linux definitely needs root (or CAP_SYS_ADMIN) to become the DRM master, so it's a bit redundant to check anything else in this case.
The libcap version is the one that should be checking these sorts of permissions, but I think they currently will fail gracefully somewhere else if you don't.

@myfreeweb @johalun Does FreeBSD have similar requirements for becoming DRM master?

Reply:
No, no requirements. Like I wrote in the log string — all you need is permissions on tty and drm devices. I have been launching rootston as a normal user.

Reply:
The problem with testing access is that `have_permissions(void)` only has access to the void :) not to variables that store which drm/tty device that was selected. I think just showing a general "make sure you have permissions" message is fine.

Reply:
This function was written purely to test if the process was able to become the DRM master, which is not device-specific in Linux.